### PR TITLE
Adding a service for clearing a costmap (except some region around the robot)

### PR DESCRIPTION
**Pull request by orduno**
## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #551  |
| Primary OS tested on | Ubuntu 16.04 |
| Robotic platform tested on | gazebo simulation of turtlebot) |

---

## Description of contribution in a few bullet points

Adding similar funcionality to [clear costmap recovery](http://wiki.ros.org/clear_costmap_recovery):

>This package provides a recovery behavior for the navigation stack that attempts to clear space by reverting the costmaps used by the navigation stack to the static map outside of a given area.

In other words, this clears everything on specific layers of a costmap, except some area around the robot. 

Instead of making it a [recovery behavior](http://wiki.ros.org/nav_core?distro=melodic#RecoveryBehavior) (as nav1 `clear_costmap_recovery` is), we move the functionality of clearing a costmap structure directly to `costmap_2d` and make it a service.

This is the first step in defining the full behavior -- see #406 

---

## Future work that may be required in bullet points

Additional steps in #406

### CODE REVIEW

**Comment by orduno** [comment]
Here's a short demo:

clear_costmap

File: nav2_costmap_2d/src/costmap_cleaner.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  double end_point_y = start_point_y + reset_distance;
+
+  int start_x, start_y, end_x, end_y;
+  costmap->worldToMapNoBounds(start_point_x, start_point_y, start_x, start_y);
+  costmap->worldToMapNoBounds(end_point_x, end_point_y, end_x, end_y);
+
+  unsigned char * grid = costmap->getCharMap();
+
+  for (int x = 0; x < static_cast<int>(costmap->getSizeInCellsX()); x++) {
+    bool isOutXrange = x<start_x || x> end_x;
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_cleaner.cpp]: You should first find the range of indices and only iterate over those, maps can be massive (think far larger than 100x100 meters), iterating over the full window for a subsection would be very costly

File: nav2_costmap_2d/src/costmap_cleaner.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    RCLCPP_ERROR(node_->get_logger(), "Cannot clean map because robot pose cannot be retrieved.");
+    return;
+  }
+
+  auto layers = costmap_->getLayeredCostmap()->getPlugins();
+
+  for (auto & layer : *layers) {
+    auto name = getLayerName(*layer);
+
+    if (any_of(
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_cleaner.cpp]: Suggestion: this is a more readability thing, if this was implemented using usual C++98 methods, this would be far easier for the typical person to understand, this takes me about 2-3 passes to understand the full intention of a pretty simple operation.

**Comment by mkhansen-intel** [comment]
I agree with @SteveMacenski - If the functionality is the same as 'Clear costmap' from the ROS implementation, even though this is a new implementation, we should keep the service and interface name as 'Clear' costmap and not 'Clean'.  If this is doing something different / better than the 'Clear' costmap function did, we need to document the difference to users.

**Comment by orduno** [comment]
I was thinking the `CostmapCleaner` would encapsulate features related to cleaning a costmap. One of which is clearing sections of it. I'd like to propose keeping the name of the class, however I'll change the name of the methods and ROS service to be `clear()` :)

File: nav2_costmap_2d/src/costmap_cleaner.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+using std::vector;
+using std::string;
+using std::shared_ptr;
+using std::any_of;
+using CleanCostmap = nav2_msgs::srv::CleanCostmap;
+
+CostmapCleaner::CostmapCleaner(rclcpp::Node::SharedPtr & node, Costmap2DROS * costmap)
+: node_(node), costmap_(costmap)
+{
+  node_->get_parameter_or_set("cleanable_layers", cleanable_layers_, {"obstacle_layer"});
```
Comment (crdelsey) [inline nav2_costmap_2d/src/costmap_cleaner.cpp]: I'd rather the clear service just called all the layers and told them to clear. Leave it up to the layer to decide whether it can be cleared or not.

Looks like costmap_ros already has a `resetLayers()` function. Why not just call that?

File: nav2_costmap_2d/src/costmap_cleaner.cpp (hunk 1 above)
Comment (crdelsey) [reply nav2_costmap_2d/src/costmap_cleaner.cpp]: @orduno Why not stick with the function that already exists `setConvexPolygonCost`

File: nav2_costmap_2d/src/costmap_cleaner.cpp [hunk 4]
```diff
@@ -27,41 +27,41 @@ using std::vector;
 using std::string;
 using std::shared_ptr;
 using std::any_of;
-using CleanCostmap = nav2_msgs::srv::CleanCostmap;
+using ClearCostmap = nav2_msgs::srv::ClearCostmap;
 
 CostmapCleaner::CostmapCleaner(rclcpp::Node::SharedPtr & node, Costmap2DROS * costmap)
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_cleaner.cpp]: Missed a spot :wink:

**Comment by crdelsey** [comment]
> I was thinking the `CostmapCleaner` would encapsulate features related to cleaning a costmap. One of which is clearing sections of it. I'd like to propose keeping the name of the class, however I'll change the name of the methods and ROS service to be `clear`

I think that is worse. Then it's not even self-consistent. I think `clear` is more commonly used terminology for what we want to do, and it is the terminology that has a precedent established for this project.

File: nav2_costmap_2d/src/costmap_cleaner.cpp (hunk 3 above)
Comment (orduno) [reply nav2_costmap_2d/src/costmap_cleaner.cpp]: Clearing/resetting with those methods clears all the layers entirely.

**Comment by orduno** [comment]
> The ROS1 equivalent consists of just this
> 
> ```
>  bool MoveBase::clearCostmapsService(std_srvs::Empty::Request &req, std_srvs::Empty::Response &resp){
>     //clear the costmaps
>     boost::unique_lock<costmap_2d::Costmap2D::mutex_t> lock_controller(*(controller_costmap_ros_->getCostmap()->getMutex()));
>     controller_costmap_ros_->resetLayers();
> 
>     boost::unique_lock<costmap_2d::Costmap2D::mutex_t> lock_planner(*(planner_costmap_ros_->getCostmap()->getMutex()));
>     planner_costmap_ros_->resetLayers();
>     return true;
> }
> ```
> We only need half that because we only have one costmap. Why not just do that?

Clearing/resetting with `restLayers()` clears all the layers entirely.

File: nav2_costmap_2d/src/costmap_cleaner.cpp (hunk 1 above)
Comment (orduno) [reply nav2_costmap_2d/src/costmap_cleaner.cpp]: I didn't provide enough info on the PR description and that caused some confusion. -- We are clearing everything except some region around the robot. 

I updated the code to use a more efficient method for clearing.

File: nav2_costmap_2d/src/costmap_cleaner.cpp (hunk 2 above)
Comment (orduno) [reply nav2_costmap_2d/src/costmap_cleaner.cpp]: I updated the code to improve the readability of this line, while still using `std::any_of`.

File: nav2_costmap_2d/src/clear_costmap_service.cpp [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  costmap->resetMapToValue(end_x, start_y, size_x, end_y, NO_INFORMATION);
+  // bottom region
+  costmap->resetMapToValue(0, end_y, size_x, size_y, NO_INFORMATION);
+
+  double ox = costmap->getOriginX(), oy = costmap->getOriginY();
+  double width = costmap->getSizeInMetersX(), height = costmap->getSizeInMetersY();
+  costmap->addExtraBounds(ox, oy, ox + width, oy + height);
+}
+
+bool ClearCostmapService::getPose(double & x, double & y) const
```
Comment (mhpanah) [inline nav2_costmap_2d/src/clear_costmap_service.cpp]: Suggestion:  Pose contains both position and orientation. I suggest to rename the method to getPosition since it only updates 2d position.

File: nav2_costmap_2d/include/nav2_costmap_2d/clear_costmap_service.hpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    const std::shared_ptr<rmw_request_id_t> request_header,
+    const std::shared_ptr<nav2_msgs::srv::ClearCostmapExceptRegion::Request> request,
+    const std::shared_ptr<nav2_msgs::srv::ClearCostmapExceptRegion::Response> response);
+
+  void clearLayerExceptRegion(
+    std::shared_ptr<CostmapLayer> & costmap, double pose_x, double pose_y, double reset_distance);
+
+  bool isClearable(const std::string & layer_name) const;
+
+  bool getPose(double & x, double & y) const;
```
Comment (mhpanah) [inline nav2_costmap_2d/include/nav2_costmap_2d/clear_costmap_service.hpp]: Why not to obtain the position x and y from `getRobotPose()` directly? What's the need for `getPose()` method.

File: nav2_costmap_2d/src/clear_costmap_service.cpp (hunk 5 above)
Comment (orduno) [reply nav2_costmap_2d/src/clear_costmap_service.cpp]: Good idea, I'll rename it.

File: nav2_costmap_2d/include/nav2_costmap_2d/clear_costmap_service.hpp (hunk 6 above)
Comment (orduno) [reply nav2_costmap_2d/include/nav2_costmap_2d/clear_costmap_service.hpp]: I just think it makes the function using it easier to read.

**Comment by crdelsey** [comment]
After talking to @orduno, I now understand that there are 3 clear_costmap type actions that exist in the ROS1 codebase.
1. [A big reset of all layers](https://github.com/ros-planning/navigation/blob/a6f3d4b49335b6de5c33044dd2123df0f491453f/move_base/src/move_base.cpp#L330-L338)
2. [Clearing all locations in a region](https://github.com/ros-planning/navigation/blob/a6f3d4b49335b6de5c33044dd2123df0f491453f/move_base/src/move_base.cpp#L275-L328)
3. [Clearing all locations outside a region](https://github.com/ros-planning/navigation/blob/a6f3d4b49335b6de5c33044dd2123df0f491453f/clear_costmap_recovery/src/clear_costmap_recovery.cpp)

This change implements 3.

File: nav2_costmap_2d/src/costmap_2d.cpp [hunk 7]
```diff
@@ -95,11 +95,17 @@ void Costmap2D::resetMaps()
...
+}
+
+void Costmap2D::resetMapToValue(
+  unsigned int x0, unsigned int y0, unsigned int xn, unsigned int yn, unsigned char value)
 {
   std::unique_lock<mutex_t> lock(*(access_));
   unsigned int len = xn - x0;
   for (unsigned int y = y0 * size_x_ + x0; y < yn * size_x_ + x0; y += size_x_) {
-    memset(costmap_ + y, default_value_, len * sizeof(unsigned char));
+    memset(costmap_ + y, value, len * sizeof(unsigned char));
```
Comment (crdelsey) [inline nav2_costmap_2d/src/costmap_2d.cpp]: I like this refactor, but I wonder if ROS1 was doing the wrong thing resetting the map to NO_INFORMATION. It kind of seems like setting it to the default_value might be more correct. Still, let's stick with ROS1 behavior unless we're fairly sure.

File: nav2_costmap_2d/src/clear_costmap_service.cpp [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+void ClearCostmapService::clearExceptRegionCallback(
+  const shared_ptr<rmw_request_id_t>/*request_header*/,
+  const shared_ptr<ClearExceptRegion::Request> request,
+  const shared_ptr<ClearExceptRegion::Response>/*response*/)
+{
+  RCLCPP_INFO(node_->get_logger(), "Received request to clear " + costmap_->getName());
+
+  if (costmap_ == nullptr) {
```
Comment (crdelsey) [inline nav2_costmap_2d/src/clear_costmap_service.cpp]: If you change the constructor to take a `Costmap2DROS & costmap` instead, then you don't need this check.

File: nav2_costmap_2d/src/clear_costmap_service.cpp [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    RCLCPP_ERROR(node_->get_logger(), "Costmap is undefined. Doing nothing.");
+    return;
+  }
+
+  RCLCPP_INFO(node_->get_logger(), "Will proceed with clearing the costmap");
```
Comment (crdelsey) [inline nav2_costmap_2d/src/clear_costmap_service.cpp]: I'd just get rid of this INFO line. The one at the start of the function is enough.

File: nav2_costmap_2d/src/clear_costmap_service.cpp [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+      auto costmap_layer = std::static_pointer_cast<CostmapLayer>(layer);
+      clearLayerExceptRegion(costmap_layer, x, y, reset_distance);
+    }
+  }
+}
+
+bool ClearCostmapService::isClearable(const string & layer_name) const
+{
+  return any_of(begin(clearable_layers_), end(clearable_layers_),
+           [&layer_name](auto l) {return l == layer_name;});
```
Comment (crdelsey) [inline nav2_costmap_2d/src/clear_costmap_service.cpp]: I'd suggest
```
auto it = find(begin(clearable_layers_), end(clearable_layers_), layer_name);
return it != end(clearable_layers_);
```
It's a bit more direct and probably easier to parse

File: nav2_costmap_2d/src/costmap_2d.cpp (hunk 7 above)
Comment (SteveMacenski) [reply nav2_costmap_2d/src/costmap_2d.cpp]: I wouldn't do this - I'd make sure to use the default value that can be set from a parameterization. In not all situations do you want a clear to result in no information. There's a parameter around called track_unknown_space which sets the default value to FREE or UNKNOWN (which you call here as NO_INFORMATION). This would be relevent for robotics applications where you want the robot to only be able to plan and control in areas that are known to be free, rather than not knowing anything about it.

For instance, I'm a walking robot (or a warehouse robot) and I clear my costmaps, I dont want to back up into unknown space if I dont have sensors back there or else I could fall down a stairwell after my costmaps are cleared. On the counter, I'm a walking or warehouse robot in a particularly sensitive area of the warehouse where it might be dangerous to go somewhere or a hospital, then I want to only go places in my map (that has been created) that I know are free based on that mapping knowledge. To stop it from going elsewhere in unknown or unexplored space (like the ER), I would set track_unknown_space to false so that free in map is free in reality and then tell my planner to disallow the robot from planning into those zones that are unmapped intentionally to not allow the robot to go anywhere unmapped or unexplored (intentionally). 

tl;dr, I'd recommend making it `default_value_` otherwise, you should go through and just completely remove the track_unknown_space logic and parameters so people dont think that will still work

File: nav2_costmap_2d/src/clear_costmap_service.cpp (hunk 8 above)
Comment (orduno) [reply nav2_costmap_2d/src/clear_costmap_service.cpp]: Good point, using a reference makes more sense in this case.

File: nav2_costmap_2d/src/clear_costmap_service.cpp (hunk 9 above)
Comment (orduno) [reply nav2_costmap_2d/src/clear_costmap_service.cpp]: Yes, even more now that we don't check for nullptr.

File: nav2_costmap_2d/src/costmap_2d.cpp (hunk 7 above)
Comment (orduno) [reply nav2_costmap_2d/src/costmap_2d.cpp]: Thank for pointing this out. I'll change it to use the `default_value_`.

File: nav2_costmap_2d/src/clear_costmap_service.cpp (hunk 10 above)
Comment (orduno) [reply nav2_costmap_2d/src/clear_costmap_service.cpp]: It's debatable which is better to use for this case, if `find`, `count` or `any_of`. 

I like `find` and `count` because we avoid using a lamba, however, I feel `any_of` express the intention more directly.

I'll change it to:
```cpp
return count(begin(clearable_layers_), end(clearable_layers_), layer_name) != 0;
```

Comment (crdelsey) [reply nav2_costmap_2d/src/clear_costmap_service.cpp]: I'd say `find` seems to be the default when you want to determine if a container contains a value, but I'll admit comparing to 0 is a touch prettier than `end(clearable_layers_)`

Find is likely to be more efficient on a large container since it can stop at the first instance instead of iterating to the end, however, in this case the list is likely to be very small so it doesn't really matter.
